### ISSUE TITLE
Implement Numeric.AD.Mode.Reverse.Double

### ISSUE THREAD

**Issue by NickSeagull** [DataHaskell/OrganizationTasks#4]

**Comment by o1lo01ol1o** [comment]
If this is done with with an eye towards back-propagation, it will need to be implemented such that it operates on tensors (or at least vectors).  This is a current limitation of the AD library.

**Comment by ocramz** [comment]
it does work out of the box for Data.Vector and all instances of Traversable IIRC

**Comment by o1lo01ol1o** [comment]
Yes, it works, but I remember reading that performance for reverse mode back-propagation would be painful for large matrices.  this may be related:  https://github.com/ekmett/ad/issues/2

When i was looking into this I considered using symbolic differentiation via an AST as a placeholder for real AD based on tensors.  These two do something similar:  http://www.scs.stanford.edu/16wi-cs240h/projects/bradbury_kathawala.pdf  They had a git hub but i'm not sure about the status of the code.

**Comment by mstksg** [comment]
It's impractical with `Data.Vector`'s Traversable instance, honestly -- mostly because any sort of linear algebra or matrix operations using `Data.Vector` is impractical at any usable scale.  At the very least, `Data.Vector.Storable` would be preferred for numeric code, but that doesn't work with *ad* because there's no `Traversable` instance.

Even if you're using `Data.Vector.Storable`,  you still probably wan't want to do actual element-by-element operations when doing matrix multiplications.  You'd want to use an FFI library for fast matrix operations like *hmatrix*.  But anything you do with *ad* naively would necessarily involve the slower element-by-element operations.

So, even if you could rewrite reverse-mode AD to work with non-traversables like `Data.Vector.Storable`, you'd still not be able to take advantage of proper linear algebra/matrix libraries :(

I'm actually implementing an experimental tensor-based backprop library myself, like in the papers mentioned earlier :)  It works at the matrix operation level, so you can take advantage of libraries like hmatrix/hblas/accelerate.

**Comment by o1lo01ol1o** [comment]
@mstksg  Is the library experimental in the sense of "pre-alpha" or "I'm not the extent to which it would work"?  Having the former working on an accelerate backend would go a long way towards gpu-based tensor optimization, even if it lagged behind cuBlas and cuDNN bindings.

**Comment by ocramz** [comment]
@mstksg I don't understand the distinction you make between elementwise operations in the Haskell runtime and the "long way around" (copy to Storable, do stuff, copy back results). Aren't the latter elementwise operations as well?

Also, I don't understand this remark:
> even if you could rewrite reverse-mode AD to work with non-traversables like Data.Vector.Storable, you'd still not be able to take advantage of proper linear algebra/matrix libraries :(

**Comment by mstksg** [comment]
@o1lo01ol1o in that it's only at the proof-of-concept stage at the moment!

@ocramz By element-wise, I mean implementing matrix multiplication with manual indexing and looping haskell vs. FFI-ing out to blas/gpu bindings.  Any "pure-haskell" matrix multiplication algorithm you could write by multiplying and adding things element-by-element will be asymptotically slower than hmatrix or other FFI bindings to libraries that perform full multiplications in C.

The only way you'd be able to use `ad` with matrices at the moment would be to implement the matrix multiplications in pure Haskell, manually indexing into matrices one-by-one and multiplying and adding elements together.  And in that case, `ad` would be able to auto-differentiate that for you ... and implement *that* using indexing into matricies one-by-one, multiplying and adding individual elements together, etc., in pure Haskell.

This limitation would be true even if you were able to implement reverse-mode AD to work with non-Traversables like `Storable` vectors.  You'd have to write your actual tensor operations in pure haskell, and `ad` would pop-back out pure-haskell derivatives/gradients/jacobian computing algorithms.

The only way to make this feasible would be to represent matrix/tensor operations as atomic operations, and use automatic differentiation to differentiate those atomic operations (into other atomic operations).  Then your FFI libraries could perform those atomic operations in an optimized way.

For example, the "derivative" of `y = A x` w.r.t. x is the outer product of x and the gradient of y, and and its derivative w.r.t. x is the gradient of y times A-transpose.  The "derivatives" here are actual matrix operations, and can be represented/computed as such using matrix FFI libraries ... instead of doing them in pure Haskell element-by-element so that the `ad` library can differentiate them...element-by-element, in pure Haskell, as well.

**Comment by SamuelSchlesinger** [comment]
@mstksg From the AD README:

"You can use a symbolic numeric type, like the one from simple-reflect to obtain symbolic derivatives:

Prelude Debug.SimpleReflect Numeric.AD> diff atanh x
recip (1 - x * x) * 1"

Using this, we can roll up a LLVM JIT and do the Theano thing and still exploit AD for the pure differentiation it gives us. Quite a few hoops were jumped to make that library pure in the Reverse mode, I don't think it would be practical to repeat the work.

**Comment by mstksg** [comment]
@SamuelSchlesinger You can use a symbolic numeric type, but those only give you symbolic derivatives at the numeric level, not at the tensor level.

You wouldn't have to repeat *all* of the work in adding tensor-level symbolic differentiation.  Just extend it :)

**Comment by SamuelSchlesinger** [comment]
@mstksg That's true, can you explain why a similar reflective trick on the functor wouldn't work for that as well? Like have a tensor AST.
